src/lib/operators/join_index.cpp

```diff
@@ -223,71 +220,132 @@ std::shared_ptr<const Table> JoinIndex::_on_execute() {
+                                      const bool track_index_matches, const bool is_semi_or_anti_join,
+                                      MultiPredicateJoinEvaluator& secondary_predicate_evaluator) {
+  auto& performance_data = static_cast<PerformanceData&>(*_performance_data);
+
+  const auto index_segment =
+      _index_input_table->get_chunk(index_chunk_id)->get_segment(_adjusted_primary_predicate.column_ids.second);
+  const auto& index_pos_list_size_pre_fallback = _index_pos_list->size();
+  for (ChunkID probe_chunk_id = ChunkID{0}; probe_chunk_id < _probe_input_table->chunk_count(); ++probe_chunk_id) {
+    const auto probe_segment =
+        _probe_input_table->get_chunk(probe_chunk_id)->get_segment(_adjusted_primary_predicate.column_ids.first);
+    JoinNestedLoop::JoinParams params{*_probe_pos_list,
+                                      *_index_pos_list,
+                                      _probe_matches[probe_chunk_id],
+                                      _index_matches[index_chunk_id],
+                                      track_probe_matches,
+                                      track_index_matches,
+                                      _mode,
+                                      _adjusted_primary_predicate.predicate_condition,
+                                      secondary_predicate_evaluator,
+                                      !is_semi_or_anti_join};
+    JoinNestedLoop::_join_two_untyped_segments(*probe_segment, *index_segment, probe_chunk_id, index_chunk_id, params);
   }
+  const auto& index_pos_list_size_post_fallback = _index_pos_list->size();
+  const auto& count_index_positions = index_pos_list_size_post_fallback - index_pos_list_size_pre_fallback;
+  std::fill_n(std::back_inserter(_index_pos_dereferenced), count_index_positions, false);
+  performance_data.chunks_scanned_without_index++;
 }
 
 // join loop that joins two segments of two columns using an iterator for the probe side,
 // and an index for the index side
 template <typename ProbeIterator>
-void JoinIndex::_join_two_segments_using_index(ProbeIterator probe_iter, ProbeIterator probe_end,
-                                               const ChunkID probe_chunk_id, const ChunkID index_chunk_id,
-                                               const std::shared_ptr<BaseIndex>& index) {
+void JoinIndex::_data_join_two_segments_using_index(ProbeIterator probe_iter, ProbeIterator probe_end,
+                                                    const ChunkID probe_chunk_id, const ChunkID index_chunk_id,
+                                                    const std::shared_ptr<BaseIndex>& index) {
+  for (; probe_iter != probe_end; ++probe_iter) {
+    const auto probe_side_position = *probe_iter;
+    const auto index_ranges = _index_ranges_for_value(probe_side_position, index);
+    for (const auto& index_range : index_ranges) {
+      _append_matches(std::get<0>(index_range), std::get<1>(index_range), probe_side_position.chunk_offset(),
```

Comment:
We should be able to use structured bindings here, right?

Reply:
I didn't know structured bindings in c++ before. Thanks for mentioning that. Yes, we can use them here.